lib/SILGen/SILGenExpr.cpp

```diff
@@ -3239,6 +3239,15 @@ static SILFunction *getOrCreateKeyPathSetter(SILGenModule &SGM,
                         std::move(subscriptIndices),
                         /*index for diags*/ nullptr);
 
+  // If the assigned value will need to be reabstracted, add a reabstraction
+  // component.
+  const auto loweredSubstType = subSGF.getLoweredType(lv.getSubstFormalType());
+  if (lv.getTypeOfRValue() != loweredSubstType.getObjectType()) {
+    // Logical components always re-abstract back to the substituted type.
+    assert(lv.isLastComponentPhysical());
+    lv.addOrigToSubstComponent(loweredSubstType);
+  }
```

Comment:
This logic is duplicated along several non-intersecting code paths. It looks like we should be checking for this whenever we finish assembling an `LValue`, which suggests some sort of an `LValueBuilder`.